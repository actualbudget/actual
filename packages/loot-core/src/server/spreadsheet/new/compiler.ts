import getSqlFields from './get-sql-fields';
import * as nodes from './nodes';
import {
  MOV,
  CALL,
  QUERY,
  UOP,
  BOP,
  REG1,
  SP,
  VAR,
  JUMPF,
  JUMPT,
  LABEL,
} from './ops';
import parse from './parser';
import generateSql from './sqlgen';

class Compiler {
  src;
  scopeName;
  binding;
  ops;
  dependencies;
  sqlDependencies;

  constructor() {
    this.ops = [];
    this.dependencies = [];
    this.sqlDependencies = [];
  }

  fail(msg, lineno, colno) {
    const lines = this.src.split('\n');

    let space = '';
    for (let i = 0; i < colno; i++) {
      space += ' ';
    }

    throw new Error(
      `[${lineno + 1}, ${colno + 1}] ${msg}:\n${lines[lineno]}\n${space}^`,
    );
  }

  resolveVariable(name) {
    if (name.indexOf('!') === -1) {
      return this.scopeName + '!' + name;
    }
    return name;
  }

  maybePushStack(node, si) {
    if (node instanceof nodes.Symbol) {
      // There's no need to push anything to the stack since it's a
      // direct variable reference. Just store the referenced variable
      // and pop the symbol operation off the stack.
      const op = this.ops.pop();
      return [si, op[1]];
    }

    this.ops.push([MOV, REG1, SP(si)]);
    return [si + 1, SP(si)];
  }

  compileLiteral(node, si) {
    this.ops.push([MOV, node.value, REG1]);
  }

  compileSymbol(node, si) {
    const resolved = this.resolveVariable(node.value);
    this.dependencies.push(resolved);
    this.ops.push([MOV, VAR(resolved), REG1]);
  }

  compileBinOp(node, si) {
    this.compile(node.left, si);
    // TODO: Get rid of all this and add a second pass which optimizes
    // the opcodes.
    let left;
    [si, left] = this.maybePushStack(node.left, si);

    this.compile(node.right, si + 1);
    this.ops.push([BOP, node.op, left, REG1]);
  }

  compileUnaryOp(node, si) {
    this.compile(node.target, si);
    this.ops.push([UOP, node.op, REG1]);
  }

  compileFunCall(node, si) {
    this.compile(node.callee, si);
    let callee;
    [si, callee] = this.maybePushStack(node.callee, si);

    const args = node.args.children.map((arg, i) => {
      this.compile(arg, si + i);
      this.ops.push([MOV, REG1, SP(si + i)]);
      return SP(si + i);
    });

    this.ops.push([CALL, callee, args]);
  }

  compileQuery(node, si) {
    let fields = getSqlFields(node.table, node.where)
      .concat(getSqlFields(node.table, node.groupby))
      .concat(...node.select.map(s => getSqlFields(node.table, s.expr)));

    const { sql, where } = generateSql(
      node.table,
      node.where,
      node.groupby,
      node.select,
    );

    // TODO: This is a hack, but I'm pretty sure we can get rid of all
    // of this. Just need to think through it.
    fields = fields.map(f => (f === '__cm.transferId' ? 'category' : f));

    // Uniquify them
    fields = [...new Set(fields)];

    this.sqlDependencies.push({ table: node.table, where, fields });
    this.ops.push([QUERY, sql, node.calculated]);
  }

  compileIf(node, si) {
    const L0 = LABEL();
    const L1 = LABEL();

    this.compile(node.cond, si);
    this.ops.push([MOV, REG1, SP(si)]);

    this.ops.push([JUMPF, SP(si), L0]);
    this.compile(node.body, si + 1);

    this.ops.push([JUMPT, SP(si), L1]);
    L0.resolve(this.ops.length - 1);
    this.compile(node.else_, si + 1);
    L1.resolve(this.ops.length - 1);
  }

  compileRoot(node, si) {
    node.children.forEach(node => {
      this.compile(node, si);
    });
  }

  compile(node, si) {
    const method = this['compile' + node.getTypeName()];
    if (!method) {
      this.fail(
        'Unknown node type: ' + node.getTypeName(),
        node.lineno,
        node.colno,
      );
    }
    return method.call(this, node, si);
  }

  compileSource(binding, scopeName, src) {
    this.src = src;
    this.scopeName = scopeName;
    this.binding = binding;

    this.compile(parse(src), 0);

    const resolvedBinding = this.resolveVariable(binding);

    if (this.ops.length !== 0) {
      this.ops.push([MOV, REG1, VAR(resolvedBinding)]);
    } else {
      this.ops.push([MOV, '', VAR(resolvedBinding)]);
    }

    return {
      ops: this.ops,
      dependencies: this.dependencies,
      sqlDependencies: this.sqlDependencies,
    };
  }
}

export function compile(src) {
  return compileBinding('result', 'generated', src);
}

export function compileBinding(binding, scopeName, src) {
  const compiler = new Compiler();
  return compiler.compileSource(binding, scopeName, src);
}
