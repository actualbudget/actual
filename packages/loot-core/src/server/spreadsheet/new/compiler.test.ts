import { compile } from './compiler';

describe('Compiler', () => {
  test('get-query', () => {
    compile(
      '=from transactions where acct.offbudget = 0 and category = null and (description.transfer_acct.offbudget = 1 or description.transfer_acct = null) calculate { count(date) }',
    );
  });

  test('basic', () => {
    let ops = compile(`
    =first(number(from transactions
       where
         date >= 20170101 and
         date <= 20170131 and
         acct.bank.name = 1
       calculate { sum(amount) }))
  `).ops;
    expect(ops).toMatchSnapshot();

    ops = compile('').ops;
    expect(ops).toMatchSnapshot();
  });

  test('parens', () => {
    let ops = compile('=(1 + 2)').ops;
    expect(ops).toMatchSnapshot();

    ops = compile('=(1232 + 2) - (3 + 4)').ops;
    expect(ops).toMatchSnapshot();
  });

  test('compiler binary ops', () => {
    let ops = compile('=foo + bar + baz + boo').ops;
    expect(ops).toMatchSnapshot();
  });

  test('compiler nested funcs', () => {
    let ops = compile('=min(0, number(-20000))').ops;
    expect(ops).toMatchSnapshot();
  });

  test('compiles boolean types', () => {
    let ops = compile('=if(true and 1) { 0 } else { 1 } ').ops;
    expect(ops).toMatchSnapshot();
  });

  test('query expressions', () => {
    let ops = compile(`
      =from transactions
         where amount > 0
         select { sum(amount) as a }
    `).ops;
    expect(ops).toMatchSnapshot();
  });

  test('query expressions with null', () => {
    let ops = compile(`
  =from transactions where acct.offbudget = 0 and category = null calculate { count(amount) }
    `).ops;
    expect(ops).toMatchSnapshot();
  });

  test('complex query expressions', () => {
    let ops = compile(`
      =from transactions groupby substr(date, 0, 7) select { substr(date, 0, 7), sum(amount) }
    `).ops;
    expect(ops).toMatchSnapshot();
  });

  test('query expressions with field remapping', () => {
    let ops = compile(`
      =from transactions where category = "50" select { id }
    `).ops;
    expect(ops).toMatchSnapshot();

    ops = compile(`
      =from transactions where category.name = "foo" select { id }
    `).ops;
    expect(ops).toMatchSnapshot();

    ops = compile(`
      =from transactions where category = "50" select { id, category.name }
    `).ops;
    expect(ops).toMatchSnapshot();
  });

  test('field dependencies', () => {
    let sqlDependencies = compile(
      '=from transactions where acct.offbudget = 0 and category = null and (description.transfer_acct.offbudget = 1 or description.transfer_acct = null) calculate { count(date) }',
    ).sqlDependencies;

    expect(sqlDependencies[0].fields).toMatchSnapshot();
  });
});
