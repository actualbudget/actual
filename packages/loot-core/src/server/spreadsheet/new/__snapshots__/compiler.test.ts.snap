// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler basic 1`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT sum(amount) FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      LEFT JOIN accounts t1 ON t1.id = transactions.acct
LEFT JOIN banks t2 ON t2.id = t1.bank
       WHERE ((((date >= 20170101) and (date <= 20170131)) and (t2.name = 1))) AND transactions.isParent = 0 AND transactions.tombstone = 0
     ",
    true,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(call),
    Object {
      "name": "generated!number",
      "type": "__var",
    },
    Array [
      Object {
        "index": 0,
        "type": "__stack",
      },
    ],
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(call),
    Object {
      "name": "generated!first",
      "type": "__var",
    },
    Array [
      Object {
        "index": 0,
        "type": "__stack",
      },
    ],
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler basic 2`] = `
Array [
  Array [
    Symbol(mov),
    "",
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler compiler binary ops 1`] = `
Array [
  Array [
    Symbol(mov),
    Object {
      "name": "generated!bar",
      "type": "__var",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "+",
    Object {
      "name": "generated!foo",
      "type": "__var",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "name": "generated!baz",
      "type": "__var",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "+",
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "name": "generated!boo",
      "type": "__var",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "+",
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler compiler nested funcs 1`] = `
Array [
  Array [
    Symbol(mov),
    0,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    -20000,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 1,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(call),
    Object {
      "name": "generated!number",
      "type": "__var",
    },
    Array [
      Object {
        "index": 1,
        "type": "__stack",
      },
    ],
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 1,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(call),
    Object {
      "name": "generated!min",
      "type": "__var",
    },
    Array [
      Object {
        "index": 0,
        "type": "__stack",
      },
      Object {
        "index": 1,
        "type": "__stack",
      },
    ],
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler compiles boolean types 1`] = `
Array [
  Array [
    Symbol(mov),
    true,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    1,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "and",
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(jumpf),
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "get": [Function],
      "resolve": [Function],
    },
  ],
  Array [
    Symbol(mov),
    0,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(jumpt),
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "get": [Function],
      "resolve": [Function],
    },
  ],
  Array [
    Symbol(mov),
    1,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler complex query expressions 1`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT substr(date,0,7), sum(amount) FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      
      WHERE transactions.isParent = 0 AND transactions.tombstone = 0
      GROUP BY substr(date,0,7)",
    false,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler field dependencies 1`] = `
Array [
  "acct",
  "category",
  "description",
  "isParent",
  "tombstone",
  "date",
]
`;

exports[`Compiler parens 1`] = `
Array [
  Array [
    Symbol(mov),
    1,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    2,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "+",
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler parens 2`] = `
Array [
  Array [
    Symbol(mov),
    1232,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    2,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "+",
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 0,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    3,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "index": 2,
      "type": "__stack",
    },
  ],
  Array [
    Symbol(mov),
    4,
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "+",
    Object {
      "index": 2,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(bop),
    "-",
    Object {
      "index": 0,
      "type": "__stack",
    },
    Object {
      "index": 1,
      "type": "__reg",
    },
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler query expressions 1`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT sum(amount) as a FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      
       WHERE ((amount > 0)) AND transactions.isParent = 0 AND transactions.tombstone = 0
     ",
    false,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler query expressions with field remapping 1`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT id FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      
       WHERE ((__cm.transferId = \\"50\\")) AND transactions.isParent = 0 AND transactions.tombstone = 0
     ",
    false,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler query expressions with field remapping 2`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT id FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      LEFT JOIN categories t1 ON __cm.transferId = t1.id
       WHERE ((t1.name = \\"foo\\")) AND transactions.isParent = 0 AND transactions.tombstone = 0
     ",
    false,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler query expressions with field remapping 3`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT id, t1.name FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      LEFT JOIN categories t1 ON __cm.transferId = t1.id
       WHERE ((__cm.transferId = \\"50\\")) AND transactions.isParent = 0 AND transactions.tombstone = 0
     ",
    false,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;

exports[`Compiler query expressions with null 1`] = `
Array [
  Array [
    Symbol(query),
    "
      SELECT count(amount) FROM transactions
      LEFT JOIN category_mapping __cm ON __cm.id = transactions.category
      LEFT JOIN accounts t1 ON t1.id = transactions.acct
       WHERE (((t1.offbudget = 0) and (__cm.transferId IS NULL))) AND transactions.isParent = 0 AND transactions.tombstone = 0
     ",
    true,
  ],
  Array [
    Symbol(mov),
    Object {
      "index": 1,
      "type": "__reg",
    },
    Object {
      "name": "generated!result",
      "type": "__var",
    },
  ],
]
`;
